crates/ide/src/annotations.rs

```diff
@@ -57,24 +57,22 @@ pub(crate) fn annotations(
                 continue;
             }
 
-            let action = runnable.action();
             let range = runnable.nav.focus_or_full_range();
 
+            // dbg_runnable should go after the run annotation, to prevent a clone we do it this way
+            let dbg_runnable = (runnable.debugee() && config.debug).then(|| Annotation {
+                range,
+                kind: AnnotationKind::Runnable { debug: true, runnable: runnable.clone() },
+            });
+
             if config.run {
                 annotations.push(Annotation {
                     range,
-
-                    // FIXME: This one allocates without reason if run is enabled, but debug is disabled
-                    kind: AnnotationKind::Runnable { debug: false, runnable: runnable.clone() },
+                    kind: AnnotationKind::Runnable { debug: false, runnable },
                 });
             }
 
-            if action.debugee && config.debug {
-                annotations.push(Annotation {
-                    range,
-                    kind: AnnotationKind::Runnable { debug: true, runnable },
-                });
-            }
+            annotations.extend(dbg_runnable);
```

Comment:
That's the right smelly smell, but the fix I think can be better. Taking a broader perspective, returning a vector with two items which differ only in one bool, but share a complex struct, feels wrong. I think the better fix would be to retern a single item with a bool indicated if the item is debuggable or not. The client can then split or not split the thing into two LSP items, depending on whether debugging is enabled. The client then can implement things like click do run and ctrl+click to debug. 

Given that `debugee()` is determined solely by `RunnableKind`, perhaps we shouldn't even have the `bool`  at all?

Let's do this refactor first, and then rebase the PR on top?

Reply:
I've been looking into this a bit the past days but I don't think you can do this on the VSCode side? As in check how the runnable/code lense has been invoked. I'd prefer merging this now and look at that problem at another point again I think.

Reply:
Sorry for being confusing -- in the context of `ide/src/lib.rs` I usually mean `handlers.rs` as the client.

Reply:
Ah okay that makes sense, I'll give that another look then that shouldn't be much to do I believe.

Reply:
Or so I say but we can't drop the bool there as otherwise we have two completely identical runnables, one spawed for debug one without, which also means moving the split out of `ide` seems tricky?...